File: node/vedadb-orm/src/connection.ts
/**
 * VedaDB ORM Connection (VedaORM class)
 *
 * Main entry point for the ORM. Manages the driver connection,
 * model registry, sessions, and transactions.
 */

import { DriverClient, DriverPool, DriverResult } from './types';
import { SchemaDefinition } from './schema';
import { Model } from './model';
import { Session } from './session';
import { RelationshipDefinition } from './relationships';
import { PopulationContext } from './population';
import { MigrationRunner } from './migration';
import { ConnectionError } from './errors';

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

export interface VedaORMOptions {
  /** A pre-created VedaDB client instance. */
  client?: DriverClient;
  /** A pre-created VedaPool instance. */
  pool?: DriverPool;
  /** Connection config for creating a client on the fly. */
  connection?: {
    host?: string;
    port?: number;
    timeout?: number;
    tls?: boolean;
    user?: string;
    password?: string;
  };
  /** Whether to use the pool for queries (default: true if pool is provided). */
  usePool?: boolean;
}

// ---------------------------------------------------------------------------
// VedaORM
// ---------------------------------------------------------------------------

export class VedaORM {
  private client: DriverClient | null = null;
  private pool: DriverPool | null = null;
  private options: VedaORMOptions;
  private _connected: boolean = false;
  private _models: Map<string, Model<any>> = new Map();
  private _schemas: Map<string, SchemaDefinition<any>> = new Map();
  private _relationships: Map<string, Record<string, RelationshipDefinition>> = new Map();
  private _migrationRunner: MigrationRunner | null = null;

  constructor(options: VedaORMOptions) {
    this.options = options;
    if (options.client) this.client = options.client;
    if (options.pool) this.pool = options.pool;
  }

  /**
   * Whether the ORM is connected.
   */
  get connected(): boolean {
    return this._connected;
  }

  /**
   * Map of registered models keyed by schema name.
   */
  get models(): Map<string, Model<any>> {
    return this._models;
  }

  // ---- Connection lifecycle -----------------------------------------------

  async connect(): Promise<void> {
    if (this._connected) return;

    if (this.pool) {
      // Pool-based: test connectivity by running a lightweight query
      try {
        await this.pool.query('SHOW TABLES;');
      } catch (err: any) {
        throw new ConnectionError(`Pool connection test failed: ${err.message}`);
      }
      this._connected = true;
      return;
    }

    if (this.client) {
      if (!this.client.connected) {
        await this.client.connect();
      }
      this._connected = true;
      return;
    }

    // Auto-create a client from connection options
    if (this.options.connection) {
      try {
        // Dynamic import of the vedadb driver
        const vedadb = require('vedadb');
        this.client = new vedadb.VedaDB(this.options.connection);
        await this.client!.connect();
        this._connected = true;
      } catch (err: any) {
        throw new ConnectionError(`Failed to connect: ${err.message}`);
      }
      return;
    }

    throw new ConnectionError('No client, pool, or connection options provided');
  }

  async disconnect(): Promise<void> {
    if (this.pool) {
      this.pool.close();
    } else if (this.client) {
      this.client.close();
    }
    this._connected = false;
  }

  // ---- Query execution (routes through pool or client) --------------------

  async query(sql: string): Promise<DriverResult> {
    if (!this._connected) {
      throw new ConnectionError('Not connected. Call connect() first.');
    }
    if (this.pool) {
      return this.pool.query(sql);
    }
    return this.client!.query(sql);
  }

  // ---- Model registration -------------------------------------------------

  register<T = any>(schema: SchemaDefinition<T>): Model<T> {
    const model = new Model<T>(schema, (sql) => this.query(sql));

    // Store schema and relationships
    this._schemas.set(schema.name, schema);
    if (!this._relationships.has(schema.name)) {
      this._relationships.set(schema.name, {});
    }

    // Set up population context for eager loading
    const populationCtx: PopulationContext = {
      executor: (sql) => this.query(sql),
      getSchema: (name) => this._schemas.get(name),
      getRelationships: (name) => this._relationships.get(name) || {},
      hydrateRows: (schemaName, result) => {
        const m = this._models.get(schemaName);
        return m ? m._fromResult(result) : result.toObjects();
      },
    };
    model.setPopulationContext(populationCtx);

    this._models.set(schema.name, model);
    return model;
  }

  /**
   * Get a registered model by schema name.
   */
  model<T = any>(name: string): Model<T> {
    const m = this._models.get(name);
    if (!m) throw new Error(`Model '${name}' is not registered`);
    return m as Model<T>;
  }

  // ---- Relationship wiring ------------------------------------------------

  /**
   * Define a relationship between two registered models.
   */
  defineRelationship(
    modelName: string,
    relationName: string,
    def: RelationshipDefinition
  ): void {
    const rels = this._relationships.get(modelName) || {};
    rels[relationName] = def;
    this._relationships.set(modelName, rels);

    const model = this._models.get(modelName);
    if (model) {
      model.addRelationship(relationName, def);
    }
  }

  // ---- Session / Transaction ----------------------------------------------

  async session(): Promise<Session> {
    if (this.pool) {
      return Session.create(() => this.pool!.acquire());
    }
    if (this.client) {
      // For single-client, wrap it
      return Session.create(() => Promise.resolve(this.client!));
    }
    throw new ConnectionError('No client or pool available for sessions');
  }

  async transaction<R>(fn: (session: Session) => Promise<R>): Promise<R> {
    if (this.pool) {
      return Session.transaction(() => this.pool!.acquire(), fn);
    }
    if (this.client) {
      return Session.transaction(() => Promise.resolve(this.client!), fn);
    }
    throw new ConnectionError('No client or pool available for transactions');
  }

  // ---- Migration runner ---------------------------------------------------

  migrations(): MigrationRunner {
    if (!this._migrationRunner) {
      this._migrationRunner = new MigrationRunner((sql) => this.query(sql));
    }
    return this._migrationRunner;
  }

  // ---- Sync all models to database ----------------------------------------

  async syncAll(): Promise<void> {
    for (const model of this._models.values()) {
      await model.sync();
    }
  }
}

File: python/vedadb_orm/connection.py
"""
VedaDB ORM — VedaORM connection class.

Central entry point that manages the connection pool, model registry,
and provides a ``session()`` context manager for transactions.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Type, TYPE_CHECKING

from .exceptions import ConnectionError, SchemaError
from .schema import Schema, to_create_sql
from .session import Session

if TYPE_CHECKING:
    from vedadb.client import Result
    from vedadb.pool import ConnectionPool
    from .model import BaseModel


class VedaORM:
    """Top-level ORM handle — wraps a ``ConnectionPool`` and a model registry.

    Usage::

        orm = VedaORM(host='localhost', port=6380)
        orm.connect()
        orm.register(User)
        users = User.where(active=True).all()
        orm.disconnect()
    """

    def __init__(
        self,
        host: str = "localhost",
        port: int = 6380,
        pool_min: int = 2,
        pool_max: int = 10,
        username: Optional[str] = None,
        password: Optional[str] = None,
        tls: bool = False,
        auto_reconnect: bool = True,
        timeout: float = 30.0,
    ) -> None:
        self.host = host
        self.port = port
        self.pool_min = pool_min
        self.pool_max = pool_max
        self.username = username
        self.password = password
        self.tls = tls
        self.auto_reconnect = auto_reconnect
        self.timeout = timeout

        self._pool: Optional["ConnectionPool"] = None
        self._models: Dict[str, Type["BaseModel"]] = {}

    # -- connection lifecycle -----------------------------------------------

    def connect(self) -> "VedaORM":
        """Create the underlying ``ConnectionPool`` and verify connectivity.

        Returns ``self`` for chaining.
        """
        try:
            from vedadb.pool import ConnectionPool
        except ImportError as exc:
            raise ConnectionError(
                "vedadb driver package is required — pip install vedadb"
            ) from exc

        self._pool = ConnectionPool(
            host=self.host,
            port=self.port,
            min_size=self.pool_min,
            max_size=self.pool_max,
            timeout=self.timeout,
        )
        return self

    def disconnect(self) -> None:
        """Close the pool and release all connections."""
        if self._pool is not None:
            self._pool.close()
            self._pool = None

    # -- raw query ----------------------------------------------------------

    def query(self, sql: str) -> "Result":
        """Execute a raw VedaQL statement through the pool.

        Acquires a connection, executes, and releases it back.
        """
        if self._pool is None:
            raise ConnectionError("Not connected — call connect() first")

        conn = self._pool.acquire()
        try:
            return conn.query(sql)
        finally:
            self._pool.release(conn)

    # -- model registry -----------------------------------------------------

    def register(self, model_cls: Type["BaseModel"], sync_schema: bool = False) -> None:
        """Register a model class with this ORM instance.

        If *sync_schema* is ``True``, the model's CREATE TABLE DDL is
        executed immediately (idempotent — uses IF NOT EXISTS semantics).
        """
        schema: Optional[Schema] = getattr(model_cls, "__schema__", None)
        if schema is None:
            raise SchemaError(
                f"{model_cls.__name__} does not define __schema__"
            )

        # Bind the model to this ORM.
        model_cls._orm_ref = self
        self._models[model_cls.__name__] = model_cls

        if sync_schema:
            ddl = to_create_sql(schema)
            # Wrap in IF NOT EXISTS by patching the first CREATE statement.
            ddl_safe = ddl.replace("CREATE TABLE ", "CREATE TABLE IF NOT EXISTS ", 1)
            ddl_safe = ddl_safe.replace(
                "CREATE COLLECTION ", "CREATE COLLECTION IF NOT EXISTS ", 1
            )
            for stmt in ddl_safe.split(";"):
                stmt = stmt.strip()
                if stmt:
                    try:
                        self.query(stmt + ";")
                    except Exception:
                        pass  # Index may already exist etc.

    @property
    def models(self) -> Dict[str, Type["BaseModel"]]:
        """Return the dict of registered model classes keyed by class name."""
        return self._models

    # -- sessions -----------------------------------------------------------

    def session(self) -> Session:
        """Return a new ``Session`` context manager backed by the pool."""
        if self._pool is None:
            raise ConnectionError("Not connected — call connect() first")
        return Session(self._pool)

    # -- pool property ------------------------------------------------------

    @property
    def pool(self) -> Optional["ConnectionPool"]:
        return self._pool

    # -- dynamic base model -------------------------------------------------

    @property
    def Model(self) -> Type["BaseModel"]:
        """Return a ``BaseModel`` subclass that is pre-bound to this ORM.

        Useful for defining models without an explicit ``register()`` call::

            orm = VedaORM().connect()
            Base = orm.Model

            class User(Base):
                __schema__ = Schema(name='users', ...)
        """
        from .model import BaseModel

        orm_ref = self

        class BoundBaseModel(BaseModel):
            @classmethod
            def _get_orm(cls) -> Optional["VedaORM"]:
                return orm_ref

        # Auto-register any subclass on creation.
        original_init_subclass = BoundBaseModel.__init_subclass__

        @classmethod  # type: ignore[misc]
        def _auto_register(cls_inner: type, **kw: Any) -> None:
            schema = getattr(cls_inner, "__schema__", None)
            if schema is not None:
                cls_inner._orm_ref = orm_ref
                orm_ref._models[cls_inner.__name__] = cls_inner  # type: ignore[assignment]

        BoundBaseModel.__init_subclass__ = _auto_register  # type: ignore[assignment]

        return BoundBaseModel
